File: 012/main.js
/**
 * @param {number} num
 * @return {string}
 */
var intToRoman = function (num) {
  let firstDigit = ['', 'I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX']
  let secondDigit = ['', 'X', 'XX', 'XXX', 'XL', 'L', 'LX', 'LXX', 'LXXX', 'XC']
  let thirdDigit = ['', 'C', 'CC', 'CCC', 'CD', 'D', 'DC', 'DCC', 'DCCC', 'CM']
  let fourthDigit = ['', 'M', 'MM', 'MMM']
  let arr = Array.from({length: 4})
  let i4 = Math.floor(num / 1000)
  if (i4 > 0) arr[0] = fourthDigit[i4]
  num = num % 1000
  let i3 = Math.floor(num / 100)
  if (i3 > 0) arr[1] = thirdDigit[i3]
  num = num % 100
  let i2 = Math.floor(num / 10)
  if (i2 > 0) arr[2] = secondDigit[i2]
  let i1 = num % 10
  if (i1 > 0) arr[3] = firstDigit[i1]
  return arr.join('')
}

console.log(intToRoman(45))
console.log(intToRoman(3999))

File: 012/main.cpp
#include <iostream>
#include <string.h>
using namespace std;

class Solution {
public:
    string intToRoman(int num) {
        string firstDigit[11] = {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"};
        string secondDigit[11] = {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"};
        string thirdDigit[11] = {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"};
        string fourthDigit[11] = {"", "M", "MM", "MMM"};
        string arr[5];
        int i4 = num / 1000;
        arr[0] = fourthDigit[i4];
        num = num % 1000;
        int i3 = num / 100;
        arr[1] = thirdDigit[i3];
        num = num % 100;
        int i2 = num / 10;
        arr[2] = secondDigit[i2];
        num = num % 10;
        int i1 = num %10;
        arr[3] = firstDigit[i1];
        string res = "";
        for (int i = 0; i < 4; i++) {
            res += arr[i];
        }
        return res;
    }
};

int main()
{
    Solution sol;
    string res = sol.intToRoman(3999);
    cout << res << endl;
    return 0;
}
